File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.swift
class Solution {
    func lengthOfLongestSubstring(_ s: String) -> Int {
        //sliding window
        var charTable: [Character: Int] = [:]
        let sArray = Array(s)
        var leftPtr = 0
        var maxLength = 0

        for (rightPtr, ch) in sArray.enumerated(){
            if charTable[ch] != nil && charTable[ch]! >= leftPtr {
                leftPtr = charTable[ch]! + 1
            } 
            charTable[ch] = rightPtr
            maxLength = max(maxLength, rightPtr - leftPtr + 1)
        }
        return maxLength
    }
}

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int l = 0;
        int r = 0;
        int maxLen = 0;
        Map<Character, Integer> seen = new HashMap<>();
        while(r < s.length() && r >= l){
            if(seen.containsKey(s.charAt(r)) && seen.get(s.charAt(r)) >= l){
                l = seen.get(s.charAt(r)) + 1;
            }
            seen.put(s.charAt(r), r);
            int len = r - l + 1;
            r++;
            maxLen = Math.max(len, maxLen);
        }
        return maxLen;
    }
}